Add subscribers for lifecycle events.

jamadden opened the pull request:
Fixes #6.

This code comes from our own internal intid utility that extended this one. The Before/After events are added to satisfy ordering needs when multiple utilities may be involved.

This introduces a fixed dependency on zope.intid and zope.keyreference. Right now this also pulls in ZODB, but hopefully zopefoundation/zope.intid#3 will be able to go in and reduce that.

I have wordsmithing to do on the docstring for the module (perhaps this is not the right place for that explanation, but it was very helpful to others in my team as we were first figuring out what the various intid utilities do and what a KeyReference is for).

jamadden committed:
Add subscribers for lifecycle events.

There is a basic example of configuration, currently in subscribers.py,
this may or may not be the best place for it. A substantial doc
reorginazation can come after #8, #9 and issue #6 (PR #10, this one) are
merged.

Fixes #6

jamadden commented:
The long docstring just got longer 😄 Despite that, I'd like to merge this as-is and focus on a documentation cleanup after the other issues (particularly #9) are dealt with.

icemac commented on CHANGES.rst:
```diff
@@ -5,6 +5,8 @@
 2.0.0 (unreleased)
 ==================
 
+- Add zope.lifecycleevent subscribers. You must include ``zubscribers.zcml`` 
```
zubscribers.zcml --> subscribers.zcml

icemac commented on src/zc/intid/__init__.py:
```diff
@@ -142,18 +142,73 @@ class IIdEvent(zope.interface.Interface):
+
+    This event is guaranteed to be the first event fired by the
+    subscriber that removes IDs. It will only be fired if at least
+    one utility will
```
This sentence misses a full stop at the end.

icemac commented on src/zc/intid/subscribers.py:
```diff
@@ -0,0 +1,192 @@
+    <include package="zc.intid" file="subscribers.zcml" />
+
+"""
+from __future__ import print_function, absolute_import, division
```
I see no usage for `print_function` and `division` in this file. What is the reason behind importing them 	
anyhow?

jamadden replied:
Thats just a part of the standard header my organization uses, which is where this code came from originally. Like the `#-*- coding: utf-8 -*-` line. It's just one way we make sure to write Python2/3 code. The imports are always there so that those features are always available.

icemac replied:
Sounds reasonable.

icemac commented on src/zc/intid/tests.py:
```diff
@@ -291,8 +349,8 @@ def test_no_zope_intid(self):
         import types
-        import sys
-        self.assertNotIn('zope.intid.interfaces', sys.modules)
+        #self.assertNotIn('zope.intid.interfaces', sys.modules)
```
Why does this line have a comment sign?

icemac commented on src/zc/intid/tests.py:
```diff
@@ -308,22 +366,198 @@ class I(zope.interface.Interface):
+        testing.setUp()
+        eventtesting.setUp()
+        traversingSetUp()
+        setHooks()
```
I think it would be easier to read and understand if you used dotted names here. Otherwise I have to scroll up to the top of the file to see where e. g. `testing` is imported from. (There are many testing.py modules in the Zope packages.)

icemac commented on src/zc/intid/tests.py:
```diff
@@ -308,22 +366,198 @@ class I(zope.interface.Interface):
+    def tearDown(self):
+        resetHooks()
+        setSite()
+        testing.tearDown()
```
Same here as in `setup()`.

icemac commented on src/zc/intid/tests.py:
```diff
@@ -308,22 +366,198 @@ class I(zope.interface.Interface):
+        objevents = self.obj_events
+
+        del events[:]
+        del objevents[:]
```
Why do you have to do this in each test? Could it be better to let `setUp` or `tearDown` do this? Or did I misunderstood something here?

jamadden replied:
It wasn't necessary in `test_addIntIdSubscriber`, but it was needed at that location in `test_removeIntIdSubscriber` because the manual registration just above generated  events.

I've simplified all of this.

icemac reviewed (changes requested):
LGTM for most of the code. See the suggestions and questions I added to the review.

jamadden commented:
Thanks. The tests are mostly copied from `zope.intid`. I did some minor cleanup, I'll see if I can do some more.

jamadden committed:
typos

jamadden committed:
Refactor tests.

Separate each main functional test into its own file. This greatly
simplifies the imports of most of them.

Also simplify the subscriber tests by removing some partially duplicated
code.

jamadden committed:
Correct wrong interface name in subscribers.zcml

Test this file by relying on it to configure the subscribers in
test_subscribers.py.
